plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx

```diff
@@ -25,6 +25,7 @@ export interface ProductCollectionConfig extends BlockVariation {
 		setPreviewState?: SetPreviewState;
 		initialPreviewState?: PreviewState;
 	};
+	usesReference?: string;
```

Comment:
## Question: Should we allow multiple values i.e. array of strings for `usesReference`?

**Point 1:**
For example, if `usesReference` is set to `['product', 'order']`, what should be the message in this case? I believe depending on the context, we will show a different message. If collection is inside Single Product template, we will show a message of product context. If collection is inside Order Confirmation template, we will show a message of order context.

**Point 2:**
Just to confirm, when we allow multiple values, it will be an OR condition. Right? This means that "one of" references is required for the collection to work properly. Right? 

**Point 3:**
We will also implement a prompt to select a product when the product context isn't available. But when there are multiple values, when should we show this prompt? I believe we should show this prompt when none of the contexts is available. Right?

Also, In the future, if we plan to implement a picker for archive context, how will it work when multiple values are allowed? Will product picker have more precedence over archive picker? Maybe I am overthinking it. We will know when the time comes? 😅

**Point 4:**

Can you think of collections that may require multiple values for `usesReference`? If not, maybe we can allow a single value for `usesReference` for now. Later, we can add support for multiple values if needed.

Reply:
> Can you think of collections that may require multiple values for usesReference?

I'm not completely certain, as this relates to the product side, but I think some key scenarios that necessitate multiple values here are the _Up-Sells_ and _Cross-Sells_ under Linked Products on the product editor. 

![Screenshot 2024-07-23 at 4 13 55 PM](https://github.com/user-attachments/assets/6c5efcb3-70e9-4dfa-be92-1cae3c88efd6)

These are "product collections" provided by core, and I understand that the intention is to use them in _both_ the **product** and **cart** contexts. This involves (a) displaying the products as expected in the product context and (b) using a combined list when in the cart, based on the cart contents.

cc @jarekmorawski  and @manospsyx to double-check

Reply:
> Point 2:
> Just to confirm, when we allow multiple values, it will be an OR condition. Right? This means that "one of" references is required for the collection to work properly. Right?

Exactly.

> Point 3: I believe we should show this prompt when none of the contexts is available. Right?

That's how I see it working.

> Point 4: Can you think of collections that may require multiple values for usesReference?

Multi-choice is a must from my POV. Collections that work with single product references, can work with multiple product references (like "cart", "order"). And that's visible in visibility matrix:

<img width="1214" alt="image" src="https://github.com/user-attachments/assets/c67aa4ec-5a78-4010-a5d5-0941823e01ec">

Reply:
Thanks @kmanijak! That list was exactly what I was looking for! 

Re **Point 2**

Yes! That's an OR.

Re **Point 3**

Here are some additional thoughts on this:
- We should only ask for the product when the collection uses a product reference and it's not globally provided, such as in single-product templates and the single-product block.
- The preview API should not be accessible for third-party developers if certain requirements are not met. It's important to note that the main checks relate to whether (a) a reference is used, (b) if this reference is available, and (c) if we can manually add it. If all three are false, then an "internal" preview system must be applied. This means that any function or initial state passed into that collection (for preview) should _not_ run or affect the editor until and unless the reference is provided completely.

Reply:
In other words, it is crucial to provide the necessary data when the `usesReference` prop requires it. The `rest_product_query` filter should only be executed, allowing 3PDs to utilize it, when the reference is completed.